Sources/CacheAdvance/CacheReader.swift

```diff
@@ -0,0 +1,130 @@
+    init(forReadingFrom file: URL, overwriteOldMessages: Bool) throws {
+        reader = try FileHandle(forReadingFrom: file)
+        self.overwriteOldMessages = overwriteOldMessages
+    }
+
+    deinit {
+        try? reader.closeHandle()
+    }
+
+    // MARK: Internal
+
+    var offsetInFileOfOldestMessage: UInt64 = 0
+    var offsetInFileAtEndOfNewestMessage: UInt64 = 0
+
+    var offsetInFile: UInt64 {
+        reader.offsetInFile
+    }
+
+    /// Returns the next encodable message, seeking to the beginning of the next message.
+    func nextEncodedMessage() throws -> Data? {
+        let startingOffset = offsetInFile
+
+        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
+            // We're at the last message.
+            return nil
+        }
+
+        switch try nextEncodedMessageSpan() {
+        case let .span(messageLength):
+            let message = try reader.readDataUp(toLength: Int(messageLength))
+            guard message.count > 0 else {
+                throw CacheAdvanceError.fileCorrupted
+            }
+
+            return message
+
+        case .emptyRead:
+            // We know the next message is at the end of the file header. Let's seek to it.
+            try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)
+
+            // We know there's a message to read now that we're at the start of the file.
+            return try nextEncodedMessage()
+
+        case .invalidFormat:
+            throw CacheAdvanceError.fileCorrupted
+        }
+    }
+
+    /// Seeks to the beginning of the oldest message in the file.
+    func seekToBeginningOfOldestMessage() throws {
+        try reader.seek(to: offsetInFileOfOldestMessage)
+    }
+
+    /// Seeks to the next message. Returns `true` when the span skipped represented a message.
+    /// When `false` is returned, it signifies that an empty read occurred.
+    @discardableResult
+    func seekToNextMessage() throws -> Bool {
+        switch try nextEncodedMessageSpan() {
+        case let .span(messageLength):
+            // There's a valid message here. Seek ahead of it.
+            try reader.seek(to: offsetInFile + UInt64(messageLength))
+            return true
+
+        case .emptyRead:
+            // We hit an empty read. Seek to the next message.
+            try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)
+            return false
+
+        case .invalidFormat:
+            throw CacheAdvanceError.fileCorrupted
+        }
+    }
+
+    // MARK: Private
+
+    /// Returns the next encoded message span, seeking to the end the span.
+    private func nextEncodedMessageSpan() throws -> NextMessageSpan {
+        let messageSizeData = try reader.readDataUp(toLength: MessageSpan.storageLength)
+
+        guard messageSizeData.count > 0 else {
+            // We haven't written anything to this file yet, or we've reached the end of the file.
+            return .emptyRead
```

Comment:
> FYI, it seems like we can now hit an `.emptyRead` regardless of the value of `overwriteOldMessages`

Indeed! I discovered during the process of writing this PR that in the old world we _also_ could hit an `emptyRead` if `overwriteOldMessages` were `true`, because of how we truncate the file before starting to overwrite old messages. Some of the tests I added were to ensure that we handled `emptyRead` properly.